keps/sig-network/2079-network-policy-port-range/README.md

```diff
@@ -0,0 +1,623 @@
+
+There are a number of user issues like [this](https://github.com/kubernetes/kubernetes/issues/67526) and [this](https://github.com/kubernetes/kubernetes/issues/93111) where users expose the need to create a policy that allow a range of ports but some specific port, or also cases that a user wants to create a policy that allows the egress to other cluster to the NodePort range (eg 32000-32768) and in this case, the rule should be created specifying each port separately, as:
+
+```
+spec:
+  egress:
+  - ports:
+    - protocol: TCP
+      port: 32000
+    - protocol: TCP
+      port: 32001
+    - protocol: TCP
+      port: 32002
+    - protocol: TCP
+      port: 32004 # Exception to port 32003
+[...]
+```
+
+So for the user, actually:
+* To allow a range of ports, each of them must be declared as an item from ``ports`` array
+* To make an exception needs a declaration of all ports but the exception
+
+Adding a new ``PortRange`` field inside the ``ports`` will allow a simpler creation of NetworkPolicy to the user.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+Add PortRange field to Ports in NetworkPolicy
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+TBD
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+In NetworkPolicy specification, inside ``NetworkPolicyPort`` object struct specify a new ``PortRange`` field:
+
+```
+// NetworkPolicyPort describes a port to allow traffic on
+type NetworkPolicyPort struct {
+    PortRange *string
+}
+```
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+As the owner of a web application I want to allow the traffic to my Pods in different ports without having to create multiple network policies. As an example, my application is exposed in ports 80, 443 and 8443 and I want to group this in a single rule
+
+#### Story 2
+
+I have another application that communicates with NodePorts of a different cluster and I want to allow the egress of the traffic only the NodePort range (eg. 30000-32767) as I don't know which port is going to be allocated in the other side, but don't want to create a rule for each of them.
+
+#### Story 3
+
+As the owner of an application, I want to allow the communication with a set of ports between 6000 and 7000 except port 6667.
+
+### Notes/Constraints/Caveats (Optional)
```

Comment:
As mentioned in one of the issues, this feature would be difficult/inefficient to implement in Open vSwitch-based network plugins (eg, openshift-sdn, ovn-kubernetes, Antrea), since OpenFlow does not have any way to specify port ranges. There may also be other network types that have difficulty implementing this as well.

Reply:
@danwinship I couldn't find the comment in the previous issues, but anyway I'm taking a look into OVS stuff to check the constraint (I believe you, just using this as an excuse to study it better).

If you have the comment can you please provide it so I can feed the document? 

Anyway I'm inserting this as a general caveat, it depends if the technology used by the CNI provider supports multiport. It came to my mind now that this might be a problem as well to eBPF based CNIs that would need to change how they populate the maps used by programs as well, so maybe it would be a need of the CNI to loop through all the ports and generate a rule for each of them.

@aojea also pointed into a [part of code](https://github.com/kubernetes/kubernetes/blob/97a13a0/pkg/security/podsecuritypolicy/provider.go#L421) that converts an existing ``HostPortRange`` field to comma-separated but it seems to me it's mostly related to PSP so wondering how this also could be re-used.

Anyway, I'll point that in the doc.

Reply:
Found it, never mind